subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/topology_aware_scheduler.go

```diff
@@ -0,0 +1,284 @@
+
+import (
+	"fmt"
+	"sort"
+)
+
+// topologyAwareScheduler can schedule a set of pods on a cluster view.
+type topologyAwareScheduler struct {
+	clusterView              CellList  // a list of node- (or lower-)level cells
+	podGpuNumbers            []int32   // GPU number of each pod to schedule
+	currentSolutionIndices   []int32   // indices of currently picked nodes (or GPUs)
+	currentLCA               CellList  // current LCA (lowest common ancestor) of currently picked GPUs
+	lowestLCASolution        CellList  // GPUs with the lowest LCA
+	lowestLCASolutionIndices []int32   // Indices of the GPUs with the lowest LCA
+	lowestLCALevel           CellLevel // level of the lowest LCA
+	// pack pods cross different priorities, or inside each priority. the former is for opportunistic
+	// pod scheduling (stay away from regular pods), because regular pods can avoid preempting opportunistic pods
+	// only among buddy cells. the latter is for intra-VC scheduling, because high-priority can avoid preemption
+	// in the whole cluster view, and hence we can pack pods with different priorities.
+	crossPriorityPack bool
+}
+
+// newTopologyAwareScheduler initializes the scheduler by extracting node-level cells
+// (lower-level if no node-level) from a chain cell list.
+func newTopologyAwareScheduler(ccl ChainCellList, crossPriorityPack bool) *topologyAwareScheduler {
+	var l CellLevel
+	for l = CellLevel(1); l <= CellLevel(len(ccl)); l++ {
+		if ccl[l][0].AtOrHigherThanNode() {
+			break
+		}
+	}
+	clusterView := make(CellList, len(ccl[l]))
+	copy(clusterView, ccl[l])
+	return &topologyAwareScheduler{
+		clusterView:       clusterView,
+		crossPriorityPack: crossPriorityPack}
+}
+
+func (t *topologyAwareScheduler) schedule(
+	podGpuNumbers map[int32]int32,
+	p CellPriority) map[int32][]CellList {
+
+	t.podGpuNumbers = []int32{}
+	for gpuNum, podNum := range podGpuNumbers {
+		for i := int32(0); i < podNum; i++ {
+			t.podGpuNumbers = append(t.podGpuNumbers, gpuNum)
+		}
+	}
+	sortGpuNumbers(t.podGpuNumbers)
+
+	// disable preemption first (reduce preemption)
+	priority := opportunisticPriority
+	t.updateClusterView(priority)
+	// try to fit the pods to a set of nodes
+	success := t.findNodesForPods(priority)
+	// enable preemption if scheduling failed
+	if !success && p > opportunisticPriority {
+		priority = p
+		t.updateClusterView(priority)
+		success = t.findNodesForPods(priority)
+	}
+	if !success {
+		return nil
+	}
+	// find GPUs inside the selected node for each pod
+	selectedNodes := make(CellList, len(t.podGpuNumbers))
+	for i := 0; i < len(t.currentSolutionIndices); i++ {
+		selectedNodes[i] = t.clusterView[t.currentSolutionIndices[i]]
+	}
+	nodeGpus := map[Cell]CellList{}
+	podPlacements := map[int32][]CellList{}
+	for podIndex := 0; podIndex < len(t.podGpuNumbers); podIndex++ {
+		gpuNumber := t.podGpuNumbers[podIndex]
+		node := selectedNodes[podIndex]
+		nodeGpus[node] = t.findGpusInNode(node, gpuNumber, priority, nodeGpus[node])
```

Comment:
>findGpusInNode [](start = 21, length = 14)

Why decouple findNodesForPods and findGpusInNode? They may leverage each other in future.

Reply:
Aware topology when findNodesForPods for packing.
Will consider in future.

---
In reply to: [332870613](https://github.com/microsoft/pai/pull/3684#discussion_r332870613) [](ancestors = 332870613)